src/SonarQube.Client/Services/SonarQubeClient.cs

```diff
@@ -0,0 +1,282 @@
+            }
+            if (requestTimeout <= TimeSpan.Zero)
+            {
+                throw new ArgumentException("Doesn't expect a zero or negative timeout.", nameof(requestTimeout));
+            }
+
+            this.httpClient = new HttpClient(messageHandler)
+            {
+                BaseAddress = connection.ServerUri,
+                Timeout = requestTimeout
+            };
+            this.httpClient.DefaultRequestHeaders.Authorization = AuthenticationHeaderFactory.Create(connection);
+        }
+
+        public void Dispose()
+        {
+            if (isDisposed)
+            {
+                return;
+            }
+
+            this.httpClient.Dispose();
+            this.isDisposed = true;
+        }
+
+        public Task<Result<ComponentResponse[]>> GetComponentsSearchProjectsAsync(ComponentRequest request,
+            CancellationToken token)
+        {
+            const string SearchProjectsAPI = "api/components/search_projects"; // Since 6.2; internal
+
+            var query = AppendQueryString(SearchProjectsAPI, "?organization={0}&p={1}&ps={2}&asc=true",
+                request.OrganizationKey, request.Page, request.PageSize); // TODO: should handle optional params
+
+            return InvokeSonarQubeApi(query, token,
+                stringResponse => JObject.Parse(stringResponse)["components"].ToObject<ComponentResponse[]>());
+        }
+
+        public Task<Result<ServerIssue[]>> GetIssuesAsync(string key, CancellationToken token)
+        {
+            const string BatchIssuesAPI = "batch/issues"; // Since 5.1; internal
+
+            var query = AppendQueryString(BatchIssuesAPI, "?key={0}", key);
+
+            return InvokeSonarQubeApi(query, token,
+                async (HttpResponseMessage response) =>
+                {
+                    var byteArray = await response.Content.ReadAsByteArrayAsync();
+                    // Protobuf for C# throws when trying to read outside of the buffer and ReadAsStreamAsync returns a non
+                    // seekable stream so we can't determine when to stop. The hack is to use an intermediate MemoryStream
+                    // so we can control when to stop reading.
+                    // Note we might want to use FileStream instead to avoid intensive memory usage.
+                    using (var stream = new MemoryStream(byteArray))
+                    {
+                        return ReadFromProtobufStream(stream, ServerIssue.Parser).ToArray();
+                    }
+                });
+        }
+
+        public Task<Result<OrganizationResponse[]>> GetOrganizationsAsync(OrganizationRequest request, CancellationToken token)
+        {
+            const string OrganizationsAPI = "api/organizations/search"; // Since 6.2; internal
+
+            var query = AppendQueryString(OrganizationsAPI, "?p={0}&ps={1}", request.Page, request.PageSize); // TODO: should handle optional params
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this 'TODO' comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)